### PR TITLE
PERF-#5596: Do not trigger index materialization for '.merge' result

### LINKED ISSUE #5596: Do not trigger index materialization for `.merge()` result
Currently, the `merge`'s logic materializes the index of the merged frames just to decide whether it should be dropped or not:
https://github.com/modin-project/modin/blob/4f78260d91042497b5d40305c3647bb00c482b14/modin/core/storage_formats/pandas/query_compiler.py#L492-L494

We can decide the same based on the `.merge()` arguments without materializing the index.

### PR DESCRIPTION
**Pull request by dchigarev**
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [ ] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [ ] signed commit with `git commit -s` 
- [x] Resolves #5596  
- [ ] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 1]
```diff
@@ -458,6 +458,13 @@ def merge(self, right, **kwargs):
 
             kwargs["sort"] = False
 
+            # Want to ensure that these are python lists
+            if left_on is not None and right_on is not None:
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:462]:
```suggestion
            if left_on and right_on:
```
fix also in other places

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 2]
```diff
@@ -458,6 +458,13 @@ def merge(self, right, **kwargs):
...
+                left_on = list(left_on) if is_list_like(left_on) else [left_on]
+                right_on = list(right_on) if is_list_like(right_on) else [right_on]
+            elif on is not None:
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:465]:
```suggestion
            elif on:
```
fix also in other places

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 3]
```diff
@@ -470,34 +477,53 @@ def map_func(left, right=right, kwargs=kwargs):
...
+            # It's cool if indexes are already materialized so we can easily check that, if not
+            # it's fine too, we can also decide that by columns, which tend to be already
+            # materialized quite often compared to the indexes.
+            keep_index = False
+            if self._modin_frame._index_cache is not None:
+                if left_on is not None and right_on is not None:
+                    keep_index = any(
+                        o in self.index.names
+                        and o in right_on
+                        and o in right.index.names
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:491]:
```suggestion
                        o in self.index.names
                        and (o in right_on
                        or o in right.index.names)
```
Should we check this way?

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 4]
```diff
@@ -470,34 +477,53 @@ def map_func(left, right=right, kwargs=kwargs):
...
+                        for o in left_on
+                    )
+                elif on is not None:
+                    keep_index = any(
+                        o in self.index.names and o in right.index.names for o in on
```
Comment (YarShev) [inline modin/core/storage_formats/pandas/query_compiler.py:496]:
```suggestion
                        o in self.index.names and (o in right_on or o in right.index.names) for o in on
```
Should we check this way?

File: modin/pandas/test/dataframe/test_join_sort.py [hunk 5]
```diff
@@ -329,6 +329,75 @@ def test_merge(test_data, test_data2):
...
+        ["col_key1", "idx_key2"],
+        ["col_key1", "idx_key3"],
+        ["idx_key1"],
+        ["idx_key2"],
+        ["idx_key3"],
+    ):
+        setup_cache()
+        res = modin_df1.merge(modin_df2, on=on)
+        ref = pandas_df1.merge(pandas_df2, on=on)
+        df_equals(res, ref)
```
Comment (YarShev) [inline modin/pandas/test/dataframe/test_join_sort.py:386]: use eval_general

**Comment by YarShev** [comment]
Is there any comparison of the changes against master?

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 1 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:462]: it behaves differently depending on whether it's None or an empty list:
```python
>>> df1.merge(df2, on=[])
IndexError: list index out of range
>>> df1.merge(df2, on=None)
  col0    col1 col0
     a  b    a    c
0    1  2    3    2
1    2  3    4    3
2    3  4    5    4
3    4  5    6    5
```

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 2 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:465]: it behaves differently depending on whether it's None or an empty list:
```python
>>> df1.merge(df2, on=[])
IndexError: list index out of range
>>> df1.merge(df2, on=None)
  col0    col1 col0
     a  b    a    c
0    1  2    3    2
1    2  3    4    3
2    3  4    5    4
3    4  5    6    5
```

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 3 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:491]: No, it's mandatory for both keys to be in their `on` argument and to be in an actual index level
```python
# 'idx_key2' exists in both indices but is specified only in a 'left_on'
>>> df1i.index.names
FrozenList(['idx_key1', 'idx_key2'])
>>> df2i.index.names
FrozenList(['idx_key2', 'idx_key3'])
>>> df1i.merge(df2i, left_on=["idx_key2"], right_on=["idx_key3"])
   idx_key3  data_col1  col_key1_x  col_key2_x  idx_key1  data_col2  col_key1_y  col_key2_y
0         4          2           4           4         1          4           3           3
1         5          3           5           5         2          3           4           4
2         6          4           6           6         3          2           5           5
# 'idx_key1' exists in both frame, in the left it's an index level and in the right it's a column name
>>> df1i.index.names
FrozenList(['idx_key1', 'idx_key2'])
>>> df2i.columns
Index(['idx_key1', 'data_col2', 'col_key1', 'col_key2'], dtype='object')
>>> df1i.merge(df2i, left_on=["idx_key1"], right_on=["idx_key1"])
   idx_key1  idx_key3  data_col1  col_key1_x  col_key2_x  data_col2  col_key1_y  col_key2_y
0         1         3         10           3           3          4           3           3
1         2         4          2           4           4          3           4           4
2         3         5          3           5           5          2           5           5
3         4         6          4           6           6         10           6           6
# 'idx_key2' exists in both frames as an index level and is specified in both 'left_on' and 'right_on'
>>> df1i.merge(df2i, left_on=["idx_key2"], right_on=["idx_key2"])
          idx_key3  data_col1  col_key1_x  col_key2_x  idx_key1  data_col2  col_key1_y  col_key2_y
idx_key2
2                3         10           3           3         1          4           3           3
3                4          2           4           4         2          3           4           4
4                5          3           5           5         3          2           5           5
5                6          4           6           6         4         10           6           6
```

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 4 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:496]: no, see [the comment above](https://github.com/modin-project/modin/pull/5619/files#r1100073715)

File: modin/pandas/test/dataframe/test_join_sort.py (hunk 5 above)
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_join_sort.py:386]: that would require much more lines to dispatch the `right` argument, the present approach looks clearer to me

Comment (YarShev) [reply modin/pandas/test/dataframe/test_join_sort.py:386]: I don't think that would require much more lines to change. Also, we don't check exceptions anyhow in these tests. Using eval_general will resolve it.

Note: we should always tent to use `eval_general`. If it requires enhancements, please create an issue for that.

Comment (dchigarev) [reply modin/pandas/test/dataframe/test_join_sort.py]: okay, found a way to make `eval_general` calls prettier:
```python
        eval_general(
            (modin_df1, modin_df2),
            (pandas_df1, pandas_df2),
            lambda dfs: dfs[0].merge(dfs[1], on=on),
        )
```
so now using it everywhere in my tests

File: modin/pandas/test/dataframe/test_join_sort.py [hunk 6]
```diff
@@ -329,6 +329,79 @@
...
+        ["col_key1", "idx_key3"],
+        ["idx_key1"],
+        ["idx_key2"],
+        ["idx_key3"],
+    ):
+        setup_cache()
+        eval_general(
+            (modin_df1, modin_df2),
+            (pandas_df1, pandas_df2),
+            lambda dfs: dfs[0].merge(dfs[1], on=on),
```
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/pandas/test/dataframe/test_join_sort.py:387]: Should we fix this?

**Comment by YarShev** [comment]
> Is there any comparison of the changes against master?

Overall, LGTM! Please also provide some time measurements.

File: modin/pandas/test/dataframe/test_join_sort.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dchigarev) [reply modin/pandas/test/dataframe/test_join_sort.py:387]: this warns us of situations like this:
```python
>>> lambdas = []
>>> for i in range(3):
...     lambdas.append(lambda: i)
...
>>> for lambd in lambdas:
...     print(lambd())
...
2
2
2
```
In our case, we're creating, executing and dropping the lambda inside a single iteration, so I would assume that this is a false warning.

**Comment by dchigarev** [comment]
@YarShev 
ASV results:
```
MODIN_TEST_DATASET_SIZE="Big" asv continuous origin/master HEAD --launch-method=spawn -b TimeMergeDefault -b TimeMerge -b TimeMergeCategoricals
```
```
       before           after         ratio
     [a4504a0e]       [2bc9380d]
      <master>       <issue_5596>
+         159±1ms         187±30ms     1.18  benchmarks.TimeMerge.time_merge_dataframe_empty_left([[500000, 20], [1000000, 10]], 'inner', True)
-       2.37±0.1s       1.82±0.04s     0.77  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[5000, 5000], [5000, 5000]], 'inner', True)
-      2.87±0.02s        2.17±0.1s     0.75  benchmarks.TimeMergeDefault.time_merge([[500000, 20], [1000000, 10]], 'inner', False)
-      2.95±0.06s       2.08±0.09s     0.71  benchmarks.TimeMergeDefault.time_merge([[500000, 20], [1000000, 10]], 'inner', True)
-       3.45±0.1s        2.42±0.1s     0.70  benchmarks.TimeMergeDefault.time_merge([[500000, 20], [1000000, 10]], 'left', True)
-      2.62±0.03s       1.83±0.03s     0.70  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[5000, 5000], [5000, 5000]], 'inner', False)
-       2.21±0.1s       1.52±0.03s     0.69  benchmarks.TimeMergeCategoricals.time_merge_categoricals([10000, 2], 'object')
-       3.09±0.1s       1.83±0.07s     0.59  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[5000, 5000], [5000, 5000]], 'left', True)
-      3.08±0.09s        1.81±0.1s     0.59  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[5000, 5000], [5000, 5000]], 'left', False)
-       2.83±0.3s       1.56±0.05s     0.55  benchmarks.TimeMergeCategoricals.time_merge_categoricals([10000, 2], 'category')
-      1.24±0.01s         522±10ms     0.42  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[500000, 20], [1000000, 10]], 'inner', False)
-      1.43±0.08s         587±40ms     0.41  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[500000, 20], [1000000, 10]], 'inner', True)
-      1.94±0.07s         748±60ms     0.38  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[500000, 20], [1000000, 10]], 'left', False)
-       2.12±0.2s         757±50ms     0.36  benchmarks.TimeMerge.time_merge_dataframe_empty_right([[500000, 20], [1000000, 10]], 'left', True)
```

**Review by YarShev — APPROVED** [review]

**Pull request merged**